spec-docs/function-types.md

```diff
@@ -0,0 +1,304 @@
+# Function Types in Kotlin on JVM
+
+## Goals
+
+* Get rid of 23 hardwired physical function classes. One of the problems with them is that they should be effectively duplicated in reflection which means a lot of physical classes in the runtime.
+* Make extension functions coercible to normal functions (with an extra parameter), so that it's possible to do `listOfStrings.map(String::length)`
```

Comment:
"coercible" usually assumes some transformation of the value representation to fit to a new type. If it's exactly the same object, just viewed as an instance of a new type, then maybe it would be better to say "implicitly convertible" or "assignable"?